src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -36,43 +33,53 @@ internal sealed class FileChangeWatcher
         /// For performance and correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block and add any work to
         /// the end of the queue.
         /// </summary>
-        private Task<IVsAsyncFileChangeEx> _taskQueue;
-        private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
-            async (precedingTask, state) =>
-            {
-                var action = (Func<IVsAsyncFileChangeEx, Task>)state;
-                await action(precedingTask.Result).ConfigureAwait(false);
-                return precedingTask.Result;
-            };
+        private readonly AsyncQueue<Func<IVsAsyncFileChangeEx, CancellationToken, Task>> _taskQueue = new();
```

Comment:
Will the delegate ever need the main thread? If so, consider  `ReentrantSemaphore` with a JTF instance instead to avoid deadlocks.

Reply:
We generally assume code never needs the main thread, until we hit a deadlock which we then go back and fix.

Reply:
The overall design here also considers any waiting on the queue to be complete to be a bug, since that then means you're blocking behind:

1. Other components operating on the file watcher service singleton which aren't in Roslyn.
2. I/O.

Even if it didn't deadlock, it's still a Bad Idea™ for performance.